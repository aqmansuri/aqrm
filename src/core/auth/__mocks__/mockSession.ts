import { Session } from '../../models';

export const mockSession: Session = {
  token: '3884915llf950',
  user: {
    id: 42,
    firstName: 'Joe',
    lastName: 'Tester',
    email: 'test@test.org',
  },
};
