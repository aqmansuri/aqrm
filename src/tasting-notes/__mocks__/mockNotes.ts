import { TastingNote } from '../../shared/models';

export const mockNotes: TastingNote[] = [
  {
    id: 4,
    brand: 'Lipton',
    name: 'Yellow Label',
    notes: 'Overly acidic, highly tannic flavor',
    rating: 4,
    teaCategoryId: 1,
  },
  {
    id: 73,
    brand: 'Bently',
    name: 'Brown Label',
    notes: 'Essentially OK',
    rating: 3,
    teaCategoryId: 2,
  },
  {
    id: 42,
    brand: 'Lipton',
    name: 'Yellow Label',
    notes: 'Overly acidic, highly tannic flavor',
    rating: 1,
    teaCategoryId: 3,
  },
];
