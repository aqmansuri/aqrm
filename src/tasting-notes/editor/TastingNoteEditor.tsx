import { useState, useEffect } from 'react';
import {
  IonButton,
  IonButtons,
  IonContent,
  IonFooter,
  IonHeader,
  IonIcon,
  IonInput,
  IonItem,
  IonLabel,
  IonList,
  IonSelect,
  IonSelectOption,
  IonTextarea,
  IonTitle,
  IonToolbar,
} from '@ionic/react';
import { useForm, Controller } from 'react-hook-form';
import { TastingNote, Tea } from '../../shared/models';
import { Rating } from '../../shared/components';
import { close } from 'ionicons/icons';
import { useTea } from '../../tea/useTea';
import { useTastingNotes } from '../useTastingNotes';

interface TastingNoteEditorProps {
  onDismiss: (opts: { refresh: boolean }) => void;
  note?: TastingNote;
}

const TastingNoteEditor: React.FC<TastingNoteEditorProps> = ({
  onDismiss,
  note = undefined,
}) => {
  const {
    handleSubmit,
    control,
    formState: { isValid },
    reset,
  } = useForm<TastingNote>({ mode: 'onChange' });
  const { getTeas } = useTea();
  const { saveNote } = useTastingNotes();
  const [teas, setTeas] = useState<Tea[]>([]);

  const save = async (data: TastingNote) => {
    if (note?.id) data.id = note.id;
    await saveNote(data);
    reset();
    onDismiss({ refresh: true });
  };

  useEffect(() => {
    (async () => {
      const teas = await getTeas();
      setTeas(teas);
    })();
  }, [getTeas]);

  return (
    <>
      <IonHeader>
        <IonToolbar>
          <IonTitle>{note ? 'Update' : 'Add New'} Tasting Note</IonTitle>
          <IonButtons slot="primary">
            <IonButton
              data-testid="cancel-button"
              onClick={() => onDismiss({ refresh: false })}
            >
              <IonIcon slot="icon-only" icon={close} />
            </IonButton>
          </IonButtons>
        </IonToolbar>
      </IonHeader>
      <IonContent className="main-content">
        <form>
          <IonList>
            <IonItem>
              <IonLabel position="floating">Brand</IonLabel>
              <Controller
                render={({ field: { onChange, value } }) => (
                  <IonInput
                    data-testid="brand-input"
                    onIonChange={e => onChange(e.detail.value!)}
                    value={value}
                  />
                )}
                control={control}
                name="brand"
                rules={{ required: true, minLength: 1 }}
                defaultValue={note?.brand || ''}
              />
            </IonItem>
            <IonItem>
              <IonLabel position="floating">Name</IonLabel>
              <Controller
                render={({ field: { onChange, value } }) => (
                  <IonInput
                    data-testid="name-input"
                    onIonChange={e => onChange(e.detail.value!)}
                    value={value}
                  />
                )}
                control={control}
                name="name"
                rules={{ required: true, minLength: 1 }}
                defaultValue={note?.name || ''}
              />
            </IonItem>
            <IonItem>
              <IonLabel>Category</IonLabel>
              <Controller
                render={({ field: { onChange, value } }) => (
                  <>
                    {teas.length && (
                      <IonSelect
                        data-testid="category-select"
                        onIonChange={e => onChange(e.detail.value!)}
                        value={value}
                      >
                        {teas.map((tea: Tea) => (
                          <IonSelectOption key={tea.id} value={tea.id}>
                            {tea.name}
                          </IonSelectOption>
                        ))}
                      </IonSelect>
                    )}
                  </>
                )}
                control={control}
                name="teaCategoryId"
                defaultValue={note?.teaCategoryId || 1}
              />
            </IonItem>
            <IonItem>
              <IonLabel>Rating</IonLabel>
              <Controller
                render={({ field: { onChange, value } }) => (
                  <Rating onRatingChange={onChange} initialRating={value} />
                )}
                control={control}
                name="rating"
                rules={{ required: true }}
                defaultValue={note?.rating || 0}
              />
            </IonItem>
            <IonItem>
              <IonLabel position="floating">Notes</IonLabel>
              <Controller
                render={({ field: { onChange, value } }) => (
                  <IonTextarea
                    data-testid="notes-input"
                    onIonChange={e => onChange(e.detail.value!)}
                    rows={5}
                    value={value}
                  />
                )}
                control={control}
                name="notes"
                rules={{ required: true, minLength: 1 }}
                defaultValue={note?.notes || ''}
              />
            </IonItem>
          </IonList>
        </form>
      </IonContent>
      <IonFooter>
        <IonToolbar>
          <IonButton
            data-testid="submit-button"
            type="submit"
            disabled={!isValid}
            expand="full"
            onClick={handleSubmit(data => save(data))}
          >
            {note ? 'Update' : 'Add'}
          </IonButton>
        </IonToolbar>
      </IonFooter>
    </>
  );
};
export default TastingNoteEditor;
