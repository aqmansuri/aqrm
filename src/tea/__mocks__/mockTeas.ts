import { Tea } from '../../shared/models';

export const expectedTeas = [
  {
    id: 1,
    name: 'Green',
    image: 'green.jpg',
    description: 'Green tea description.',
  },
  {
    id: 2,
    name: 'Black',
    image: 'black.jpg',
    description: 'Black tea description.',
  },
  {
    id: 3,
    name: 'Herbal',
    image: 'herbal.jpg',
    description: 'Herbal Infusion description.',
  },
  {
    id: 4,
    name: 'Oolong',
    image: 'oolong.jpg',
    description: 'Oolong tea description.',
  },
  {
    id: 5,
    name: 'Dark',
    image: 'dark.jpg',
    description: 'Dark tea description.',
  },
  {
    id: 6,
    name: 'Puer',
    image: 'puer.jpg',
    description: 'Puer tea description.',
  },
  {
    id: 7,
    name: 'White',
    image: 'white.jpg',
    description: 'White tea description.',
  },
  {
    id: 8,
    name: 'Yellow',
    image: 'yellow.jpg',
    description: 'Yellow tea description.',
  },
];

export const resultTeas = () => {
  return expectedTeas.map((t: Tea) => {
    const tea = { ...t };
    // @ts-ignore
    delete tea.image;
    return tea;
  });
};
