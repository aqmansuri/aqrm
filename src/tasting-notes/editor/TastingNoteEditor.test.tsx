import { render, waitFor } from '@testing-library/react';
import { ionFireEvent as fireEvent } from '@ionic/react-test-utils';
import TastingNoteEditor from './TastingNoteEditor';
import { mockNotes } from '../__mocks__/mockNotes';
import { resultTeas } from '../../tea/__mocks__/mockTeas';

const mockTeas = resultTeas();

jest.mock('../../tea/useTea', () => ({
  useTea: () => ({
    getTeas: jest.fn(() => Promise.resolve(mockTeas)),
  }),
}));

let mockSaveNote = jest.fn(() => Promise.resolve());

jest.mock('../useTastingNotes', () => ({
  useTastingNotes: () => ({
    saveNote: mockSaveNote,
  }),
}));

describe('<TastingNoteEditor />', () => {
  let component: any;
  let mockDismiss = jest.fn();

  beforeEach(() => {
    component = <TastingNoteEditor onDismiss={mockDismiss} />;
    mockSaveNote = jest.fn(() => Promise.resolve());
  });

  describe('initialization', () => {
    it('binds the tea select', async () => {
      const { getByTestId } = render(component);
      const options = await waitFor(() => getByTestId(/category-select/));
      expect(options.children.length).toEqual(8);
      expect(options.children[0].textContent).toEqual('Green');
      expect(options.children[1].textContent).toEqual('Black');
    });
  });

  describe('cancel button', () => {
    it('calls the dismiss function', async () => {
      const { getByTestId } = render(component);
      const button = await waitFor(
        () => getByTestId(/cancel-button/) as HTMLIonButtonElement,
      );
      fireEvent.click(button);
      expect(mockDismiss).toHaveBeenCalledTimes(1);
    });
  });

  describe('save', () => {
    it('renders consistently', async () => {
      const { asFragment } = render(component);
      await waitFor(() => expect(asFragment()).toMatchSnapshot());
    });

    it('has the add title', async () => {
      const { container } = render(component);
      await waitFor(() =>
        expect(container).toHaveTextContent(/Add New Tasting Note/),
      );
    });

    it('has the add button label', async () => {
      const { getByTestId } = render(component);
      const submit = await waitFor(() => getByTestId(/submit-button/));
      expect(submit.textContent).toEqual('Add');
    });

    it('saves the note', async () => {
      const expected = { ...mockNotes[0] };
      // @ts-ignore
      delete expected.id;
      const { getByTestId } = render(component);
      const brand = await waitFor(() => getByTestId(/brand-input/));
      const name = await waitFor(() => getByTestId(/name-input/));
      const rating = await waitFor(() => getByTestId(/Rate 4 stars/));
      const notes = await waitFor(() => getByTestId(/notes-input/));
      const submit = await waitFor(() => getByTestId(/submit-button/));

      await waitFor(() => fireEvent.ionChange(brand, mockNotes[0].brand));
      await waitFor(() => fireEvent.ionChange(name, mockNotes[0].name));
      await waitFor(() => fireEvent.click(rating));
      await waitFor(() => fireEvent.ionChange(notes, mockNotes[0].notes));
      await waitFor(() => fireEvent.click(submit));

      expect(mockSaveNote).toHaveBeenCalledWith(expected);
      expect(mockSaveNote).toHaveBeenCalledTimes(1);
    });
  });

  describe('update', () => {
    beforeEach(() => {
      component = (
        <TastingNoteEditor onDismiss={mockDismiss} note={mockNotes[0]} />
      );
      mockSaveNote = jest.fn(() => Promise.resolve());
    });

    it('renders consistently', async () => {
      const { asFragment } = render(component);
      await waitFor(() => expect(asFragment()).toMatchSnapshot());
    });

    it('has the update title', async () => {
      const { container } = render(component);
      await waitFor(() =>
        expect(container).toHaveTextContent(/Update Tasting Note/),
      );
    });

    it('has the update button label', async () => {
      const { getByTestId } = render(component);
      const submit = await waitFor(() => getByTestId(/submit-button/));
      expect(submit.textContent).toEqual('Update');
    });

    it('sets the properties', async () => {
      const { getByTestId } = render(component);
      const brand = await waitFor(() => getByTestId(/brand-input/));
      const name = await waitFor(() => getByTestId(/name-input/));
      const notes = await waitFor(() => getByTestId(/notes-input/));
      expect(brand.getAttribute('value')).toEqual(mockNotes[0].brand);
      expect(name.getAttribute('value')).toEqual(mockNotes[0].name);
      expect(notes.getAttribute('value')).toEqual(mockNotes[0].notes);
    });

    it('updates the data', async () => {
      const expected = { ...mockNotes[0] };
      expected.notes = "It's not good";
      expected.rating = 1;
      const { getByTestId } = render(component);
      const brand = await waitFor(() => getByTestId(/brand-input/));
      const name = await waitFor(() => getByTestId(/name-input/));
      const rating = await waitFor(() => getByTestId(/Rate 1 stars/));
      const notes = await waitFor(() => getByTestId(/notes-input/));
      const submit = await waitFor(() => getByTestId(/submit-button/));

      await waitFor(() => fireEvent.ionChange(brand, mockNotes[0].brand));
      await waitFor(() => fireEvent.ionChange(name, mockNotes[0].name));
      await waitFor(() => fireEvent.click(rating));
      await waitFor(() => fireEvent.ionChange(notes, expected.notes));
      await waitFor(() => fireEvent.click(submit));

      expect(mockSaveNote).toHaveBeenCalledWith(expected);
      expect(mockSaveNote).toHaveBeenCalledTimes(1);
    });
  });

  afterEach(() => jest.restoreAllMocks());
});
