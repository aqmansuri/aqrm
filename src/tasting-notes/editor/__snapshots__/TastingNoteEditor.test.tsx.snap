// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TastingNoteEditor /> save renders consistently 1`] = `
<DocumentFragment>
  <ion-header>
    <ion-toolbar>
      <ion-title>
        Add New Tasting Note
      </ion-title>
      <ion-buttons
        slot="primary"
      >
        <ion-button
          data-testid="cancel-button"
        >
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Close</title><path d='M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z'/></svg>"
            slot="icon-only"
          />
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header>
  <ion-content>
    <form>
      <ion-list>
        <ion-item>
          <ion-label
            position="floating"
          >
            Brand
          </ion-label>
          <ion-input
            data-testid="brand-input"
            value=""
          />
        </ion-item>
        <ion-item>
          <ion-label
            position="floating"
          >
            Name
          </ion-label>
          <ion-input
            data-testid="name-input"
            value=""
          />
        </ion-item>
        <ion-item>
          <ion-label>
            Category
          </ion-label>
          0
        </ion-item>
        <ion-item>
          <ion-label>
            Rating
          </ion-label>
          <div
            class="rating"
            style="opacity: 1;"
          >
            <ion-icon
              data-testid="Rate 1 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            />
            <ion-icon
              data-testid="Rate 2 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            />
            <ion-icon
              data-testid="Rate 3 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            />
            <ion-icon
              data-testid="Rate 4 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            />
            <ion-icon
              data-testid="Rate 5 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            />
          </div>
        </ion-item>
        <ion-item>
          <ion-label
            position="floating"
          >
            Notes
          </ion-label>
          <ion-textarea
            data-testid="notes-input"
            rows="5"
            value=""
          />
        </ion-item>
      </ion-list>
    </form>
  </ion-content>
  <ion-footer>
    <ion-toolbar>
      <ion-button
        data-testid="submit-button"
        disabled="false"
        expand="full"
        type="submit"
      >
        Add
      </ion-button>
    </ion-toolbar>
  </ion-footer>
</DocumentFragment>
`;

exports[`<TastingNoteEditor /> update renders consistently 1`] = `
<DocumentFragment>
  <ion-header>
    <ion-toolbar>
      <ion-title>
        Update Tasting Note
      </ion-title>
      <ion-buttons
        slot="primary"
      >
        <ion-button
          data-testid="cancel-button"
        >
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Close</title><path d='M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z'/></svg>"
            slot="icon-only"
          />
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header>
  <ion-content>
    <form>
      <ion-list>
        <ion-item>
          <ion-label
            position="floating"
          >
            Brand
          </ion-label>
          <ion-input
            data-testid="brand-input"
            value="Lipton"
          />
        </ion-item>
        <ion-item>
          <ion-label
            position="floating"
          >
            Name
          </ion-label>
          <ion-input
            data-testid="name-input"
            value="Yellow Label"
          />
        </ion-item>
        <ion-item>
          <ion-label>
            Category
          </ion-label>
          0
        </ion-item>
        <ion-item>
          <ion-label>
            Rating
          </ion-label>
          <div
            class="rating"
            style="opacity: 1;"
          >
            <ion-icon
              data-testid="Rate 1 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z'/></svg>"
            />
            <ion-icon
              data-testid="Rate 2 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z'/></svg>"
            />
            <ion-icon
              data-testid="Rate 3 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z'/></svg>"
            />
            <ion-icon
              data-testid="Rate 4 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z'/></svg>"
            />
            <ion-icon
              data-testid="Rate 5 stars"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Star</title><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            />
          </div>
        </ion-item>
        <ion-item>
          <ion-label
            position="floating"
          >
            Notes
          </ion-label>
          <ion-textarea
            data-testid="notes-input"
            rows="5"
            value="Overly acidic, highly tannic flavor"
          />
        </ion-item>
      </ion-list>
    </form>
  </ion-content>
  <ion-footer>
    <ion-toolbar>
      <ion-button
        data-testid="submit-button"
        disabled="false"
        expand="full"
        type="submit"
      >
        Update
      </ion-button>
    </ion-toolbar>
  </ion-footer>
</DocumentFragment>
`;
